File: packages/shared/src/schemas/primitives.ts
import { z } from "zod";

// Kept in lockstep with packages/shared/src/fii_shared/primitives.py.

export const Confidence = z.enum(["low", "medium", "high"]);
export type Confidence = z.infer<typeof Confidence>;

export const Trend3 = z.enum(["improving", "stable", "deteriorating"]);
export type Trend3 = z.infer<typeof Trend3>;

export const SourceType = z.enum([
  "sec_filing",
  "fmp_fundamental",
  "polygon_price",
  "fred_series",
  "finnhub_news",
  "calculated",
]);
export type SourceType = z.infer<typeof SourceType>;

// ISO date (YYYY-MM-DD) and ISO datetime strings — Python emits these verbatim.
const IsoDate = z.string().regex(/^\d{4}-\d{2}-\d{2}$/);
const IsoDateTime = z.string().datetime({ offset: true });

export const SourceRef = z.object({
  source_type: SourceType,
  source_id: z.string().min(1),
  section: z.string().nullable().optional(),
  retrieved_at: IsoDateTime,
  url: z.string().url().nullable().optional(),
});
export type SourceRef = z.infer<typeof SourceRef>;

export const CitedClaim = z.object({
  claim: z.string().min(1).max(2000),
  sources: z.array(SourceRef).min(1),
  confidence: Confidence,
});
export type CitedClaim = z.infer<typeof CitedClaim>;

export const CitedNumber = z.object({
  value: z.number(),
  unit: z.string().min(1).max(32),
  as_of: IsoDate,
  source: SourceRef, // required — no nullable/optional
});
export type CitedNumber = z.infer<typeof CitedNumber>;

// Sub-types shared across specialists.

export const CompParable = z.object({
  peer_symbol: z.string().min(1).max(10),
  ev_ebitda: z.number().nullable().optional(),
  pe_ratio: z.number().nullable().optional(),
});
export type CompParable = z.infer<typeof CompParable>;

export const InsiderMove = z.object({
  name: z.string().min(1),
  role: z.string().nullable().optional(),
  action: z.enum(["buy", "sell", "grant", "exercise", "other"]),
  shares: z.number(),
  dollar_value: z.number(),
  date: IsoDate,
});
export type InsiderMove = z.infer<typeof InsiderMove>;

const ForceLevel = z.enum(["low", "medium", "high"]);
export const FiveForces = z.object({
  threat_new_entrants: ForceLevel,
  bargaining_buyers: ForceLevel,
  bargaining_suppliers: ForceLevel,
  threat_substitutes: ForceLevel,
  competitive_rivalry: ForceLevel,
});
export type FiveForces = z.infer<typeof FiveForces>;

// Re-export helpers used by the parity test.
export { IsoDate, IsoDateTime };

File: packages/shared/src/fii_shared/primitives.py
"""Citation and primitive types used by every specialist output.

These are the contract pieces that prevent hallucination:
- SourceRef — identifies where a claim comes from
- CitedClaim — a textual claim plus its sources
- CitedNumber — a numeric value that MUST carry a source

Numbers outside a CitedNumber wrapper are allowed only for agent-computed metrics
(e.g. RSI, realized vol) where the underlying series IS the source.
"""

from __future__ import annotations

from datetime import date, datetime
from enum import StrEnum
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class Confidence(StrEnum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class Trend3(StrEnum):
    IMPROVING = "improving"
    STABLE = "stable"
    DETERIORATING = "deteriorating"


class SourceType(StrEnum):
    SEC_FILING = "sec_filing"
    FMP_FUNDAMENTAL = "fmp_fundamental"
    POLYGON_PRICE = "polygon_price"
    FRED_SERIES = "fred_series"
    FINNHUB_NEWS = "finnhub_news"
    CALCULATED = "calculated"


class SourceRef(BaseModel):
    """Where a claim comes from. Every factual assertion an agent makes must carry one."""

    model_config = ConfigDict(frozen=True, str_strip_whitespace=True)

    source_type: SourceType
    source_id: str = Field(
        min_length=1,
        description="Filing accession, FMP endpoint, FRED series ID, news URL, or formula name",
    )
    section: str | None = Field(
        default=None, description="e.g. 'Item 1A Risk Factors' or 'ratios.debtEquityRatio'"
    )
    retrieved_at: datetime
    url: HttpUrl | None = None


class CitedClaim(BaseModel):
    """A textual claim with at least one source. Used for qualitative evidence."""

    model_config = ConfigDict(str_strip_whitespace=True)

    claim: str = Field(min_length=1, max_length=2000)
    sources: list[SourceRef] = Field(min_length=1)
    confidence: Confidence


class CitedNumber(BaseModel):
    """A numeric value bound to exactly one source. `source` is non-optional — if JSON
    supplies null, Pydantic rejects. This is the primary defense against made-up numbers.
    """

    model_config = ConfigDict(str_strip_whitespace=True)

    value: float
    unit: str = Field(
        min_length=1,
        max_length=32,
        description="'USD', 'percent', 'ratio', 'count', 'months', etc.",
    )
    as_of: date
    source: SourceRef  # intentionally required, no | None


# --- Shared sub-types used across specialists --------------------------------------------


class CompParable(BaseModel):
    """One row of the valuation comparable-multiples table."""

    peer_symbol: str = Field(min_length=1, max_length=10)
    ev_ebitda: float | None = None
    pe_ratio: float | None = None


class InsiderMove(BaseModel):
    name: str = Field(min_length=1)
    role: str | None = None
    action: Literal["buy", "sell", "grant", "exercise", "other"]
    shares: float
    dollar_value: float
    date: date


class FiveForces(BaseModel):
    threat_new_entrants: Literal["low", "medium", "high"]
    bargaining_buyers: Literal["low", "medium", "high"]
    bargaining_suppliers: Literal["low", "medium", "high"]
    threat_substitutes: Literal["low", "medium", "high"]
    competitive_rivalry: Literal["low", "medium", "high"]
